python/ray/data/_internal/datasource/parquet_datasink.py

```diff
@@ -75,10 +75,12 @@ def write(
 
         def write_blocks_to_path():
             with self.open_output_stream(write_path) as file:
-                schema = BlockAccessor.for_block(blocks[0]).to_arrow().schema
+                tables = [BlockAccessor.for_block(block).to_arrow() for block in blocks]
+                schema = self._try_merge_nullable_fields(tables)
                 with pq.ParquetWriter(file, schema, **write_kwargs) as writer:
-                    for block in blocks:
-                        table = BlockAccessor.for_block(block).to_arrow()
+                    for table in tables:
+                        if not table.schema.equals(schema):
+                            table = table.cast(schema)
```

Comment:
What happens if we don't explicitly cast the tables?

Reply:
The table would still have a mismatch schema. i.e. 
`table.schema.equals(schema)` in this case would still be false.

Reply:
Gotcha. Wasn't sure if PyArrow would implicitly cast tables to match the specified schema under-the-hood

Reply:
Yeah, it doesn't do the casting because there's check on the schema equality here: 

https://github.com/apache/arrow/blob/main/python/pyarrow/parquet/core.py#L1110-L1114